### PR TITLE
CDAP-1697: Explore Service commits transaction even when operation fails

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-1697
http://builds.cask.co/browse/CDAP-DUT2481

### CODE REVIEW

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java [hunk 1]
```diff
@@ -334,19 +333,20 @@ public QueryHandle getColumns(String catalog, String schemaPattern, String table
...
       Map<String, String> sessionConf = startSession();
+
       try {
         sessionHandle = openHiveSession(sessionConf);
 
         String database = getHiveDatabase(schemaPattern);
         operationHandle = cliService.getColumns(sessionHandle, catalog, database,
-                                                tableNamePattern, columnNamePattern);
-        QueryHandle handle = saveOperationInfo(operationHandle, sessionHandle, sessionConf, "", database);
+                                                                tableNamePattern, columnNamePattern);
```
Comment (anwar6953) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java]: why indent

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java [hunk 2]
```diff
@@ -488,18 +492,19 @@ public QueryHandle getTables(String catalog, String schemaPattern, String tableN
...
       OperationHandle operationHandle = null;
       Map<String, String> sessionConf = startSession();
+
       try {
         sessionHandle = openHiveSession(sessionConf);
         String database = getHiveDatabase(schemaPattern);
         operationHandle = cliService.getTables(sessionHandle, catalog, database,
-                                               tableNamePattern, tableTypes);
-        QueryHandle handle = saveOperationInfo(operationHandle, sessionHandle, sessionConf, "", database);
+                                                               tableNamePattern, tableTypes);
```
Comment (anwar6953) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java]: why indent

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/InactiveOperationInfo.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package co.cask.cdap.explore.service.hive;
+
+import co.cask.cdap.proto.ColumnDesc;
+import co.cask.cdap.proto.QueryStatus;
+
+import java.util.List;
+
+final class InactiveOperationInfo extends OperationInfo {
```
Comment (sreevatsanraman) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/InactiveOperationInfo.java]: javadocs

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java [hunk 4]
```diff
@@ -1240,18 +1258,43 @@ protected OperationHandle getOperationHandle(QueryHandle handle) throws ExploreE
     return getOperationInfo(handle).getOperationHandle();
   }
 
+  protected QueryStatus fetchStatus(OperationInfo operationInfo)
```
Comment (poornachandra) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java]: `fetchStatus` should be a public API.

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceInvalidateTxTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    // check that BaseHiveExploreService invalidates read-write transaction that failed
+    Assert.assertEquals(0, transactionManager.getCurrentState().getInvalid().size());
+    try {
+      waitForCompletionStatus(exploreService.execute(NAMESPACE_ID, "select * from dataset_nonexistent"),
+                              5, TimeUnit.SECONDS, 10);
+      Assert.fail("Expected HiveSQLException");
+    } catch (HiveSQLException e) {
+      // Expected
+    }
+    Assert.assertEquals(1, transactionManager.getCurrentState().getInvalid().size());
```
Comment (poornachandra) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceInvalidateTxTest.java]: Can you also add a test which does not invalidate a read-only operation on failure?

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/InactiveOperationInfo.java (hunk 3 above)
Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/InactiveOperationInfo.java]: This isn't a public class. Are javadocs still necessary?

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java (hunk 4 above)
Comment (jwang47) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java]: `getStatus` is the public API whereas this one is internal, to be overridden by subclasses to support various versions of Hive.

File: cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceInvalidateTxTest.java (hunk 5 above)
Comment (jwang47) [inline cdap-explore/src/test/java/co/cask/cdap/explore/service/HiveExploreServiceInvalidateTxTest.java]: Added.

**Pull request merged**